pkg/volume/iscsi/iscsi_util.go

```diff
@@ -113,33 +114,52 @@ func (util *ISCSIUtil) AttachDisk(b iscsiDiskMounter) error {
-		return errors.New(fmt.Sprintf("Could not parse iface file for %s", b.iface))
-	} else if iscsiTransport == "tcp" {
-		devicePath = strings.Join([]string{"/dev/disk/by-path/ip", b.portal, "iscsi", b.iqn, "lun", b.lun}, "-")
-	} else {
-		devicePath = strings.Join([]string{"/dev/disk/by-path/pci", "*", "ip", b.portal, "iscsi", b.iqn, "lun", b.lun}, "-")
-	}
-	exist := waitForPathToExist(devicePath, 1, iscsiTransport)
-	if exist == false {
-		// discover iscsi target
-		out, err := b.plugin.execCommand("iscsiadm", []string{"-m", "discovery", "-t", "sendtargets", "-p", b.portal, "-I", b.iface})
-		if err != nil {
-			glog.Errorf("iscsi: failed to sendtargets to portal %s error: %s", b.portal, string(out))
-			return err
-		}
-		// login to iscsi target
-		out, err = b.plugin.execCommand("iscsiadm", []string{"-m", "node", "-p", b.portal, "-T", b.iqn, "-I", b.iface, "--login"})
-		if err != nil {
-			glog.Errorf("iscsi: failed to attach disk:Error: %s (%v)", string(out), err)
-			return err
+	bkpPortal := append(b.bkpPortal, b.portal)
+	for _, tp := range bkpPortal {
+		if iscsiTransport == "" {
+			glog.Errorf("iscsi: could not find transport name in iface %s", b.iface)
+			return errors.New(fmt.Sprintf("Could not parse iface file for %s", b.iface))
+		} else if iscsiTransport == "tcp" {
+			devicePath = strings.Join([]string{"/dev/disk/by-path/ip", tp, "iscsi", b.iqn, "lun", b.lun}, "-")
+		} else {
+			devicePath = strings.Join([]string{"/dev/disk/by-path/pci", "*", "ip", tp, "iscsi", b.iqn, "lun", b.lun}, "-")
 		}
-		exist = waitForPathToExist(devicePath, 10, iscsiTransport)
-		if !exist {
-			return errors.New("Could not attach disk: Timeout after 10s")
+		exist := waitForPathToExist(devicePath, 1, iscsiTransport)
+		if exist == false {
+			// discover iscsi target
+			out, err := b.plugin.execCommand("iscsiadm", []string{"-m", "discovery", "-t", "sendtargets", "-p", tp, "-I", b.iface})
+			if err != nil {
+				glog.Errorf("iscsi: failed to sendtargets to portal %s error: %s", tp, string(out))
+				continue
+			}
+			// login to iscsi target
+			out, err = b.plugin.execCommand("iscsiadm", []string{"-m", "node", "-p", tp, "-T", b.iqn, "-I", b.iface, "--login"})
+			if err != nil {
+				glog.Errorf("iscsi: failed to attach disk:Error: %s (%v)", string(out), err)
+				continue
+			}
+			exist = waitForPathToExist(devicePath, 10, iscsiTransport)
+			if !exist {
+				glog.Errorf("Could not attach disk: Timeout after 10s")
+			} else {
+				devicePaths = append(devicePaths, devicePath)
+			}
+		} else {
+			glog.Infof("iscsi: devicepath (%s) exist", devicePath)
+			devicePaths = append(devicePaths, devicePath)
+			reqExist = exist
 		}
+
+	}
+
+	if !reqExist {
+		glog.Errorf("iscsi: failed to get any path for iscsi disk")
+		return errors.New("failed to get any path for iscsi disk")
 	}
+
+	//Make sure we use a valid devicepath to find mpio device.
+	devicePath = devicePaths[0]
```

Comment:
devicePath array entry will not be empty. The devicepath would be  filled with a valid path in any case. It shouldnt be empty path as we are populating the array by validating the path.